### PR TITLE
generalize closure api in Lightning

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Draft for a general API for handling closures in Lightning. 
Motivation: Depending on how involved a loop customization is, sooner or later a user has to deal with closures at some point. This is quite difficult in Lightning currently due to a high entanglement with gradient accumulation, manual backward, precision etc. By extracting out a high level api for building closures, we enable customization while providing a clean implementation for manual optimization etc. 

DRAFT.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/closure.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    closure_loss: Optional[Tensor]
+    result_collection: Optional[ResultCollection]
+
+    @property
+    def loss(self) -> Optional[Tensor]:
+        if self.closure_loss is not None:
+            return self.closure_loss.detach().clone()
+
+
+class Closure(ABC):
```
Comment (justusschock) [inline pytorch_lightning/loops/closure.py:23]: Just for the naming: IMO this should be the `LightningClosure` (if having one at all), this would be consistent with `LightningEnums`, which is also the base there.

File: pytorch_lightning/loops/closure.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result = None
+
+    @property
+    def result(self) -> ClosureResult:
+        """The cached result from the last time the closure was called."""
+        return self._result
```
Comment (justusschock) [inline pytorch_lightning/loops/closure.py:42]: Do we really want a reference to this? Usually we only need to create the closure once and can reuse it. This might introduce some unwanted longtime references...

File: pytorch_lightning/loops/closure.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        """Implements the behavior of the closure once it is getting called."""
+        pass
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
+        self._result = self.closure(*args, **kwargs)
+        if self._result is not None:
+            return self._result.loss
+
+
+class LightningClosure(Closure):
```
Comment (justusschock) [inline pytorch_lightning/loops/closure.py:55]: Naming (see above): DefaultLightningClosure

**Review by justusschock — COMMENTED** [review]
One comment on memory and two really minor ones on naming :)

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 4]
```diff
@@ -309,18 +329,16 @@ def _training_step(
...
+            if result_collection is None:
                 return
 
         closure_loss = None
-        loss = None
         if self.trainer.lightning_module.automatic_optimization:
             # accumulate loss. if accumulate_grad_batches==1, no effect
-            closure_loss = training_step_output.minimize / self.trainer.accumulate_grad_batches
+            closure_loss = result_collection.minimize / self.trainer.accumulate_grad_batches
             # the loss will get scaled for amp. avoid any modifications to it
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:340]:
```suggestion
```

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 5]
```diff
@@ -211,37 +213,55 @@ def _run_optimization(
...
+        def step_fn():
+            return self._training_step(split_batch, batch_idx, opt_idx, hiddens)
+
+        return step_fn
+
+    def _make_backward_fn(self, batch_idx, optimizer, opt_idx):
+        def backward_fn(loss: Tensor):
+            self.backward(loss, optimizer, opt_idx)
+
+            loss_val = loss.detach().clone()
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:236]:
```suggestion
            # the loss will get scaled for amp. avoid any modifications to it
            loss_val = loss.detach().clone()
```

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -211,37 +213,55 @@ def _run_optimization(
...
-        opt_idx: int,
-        optimizer: Optimizer,
-        hiddens: Tensor,
-        return_result: AttributeDict,
-    ) -> Optional[Tensor]:
-        """Closure for training step and backward
+    def _make_closure(self, split_batch, batch_idx, opt_idx, optimizer, hiddens):
+        step_fn = self._make_step_fn(split_batch, batch_idx, opt_idx, hiddens)
+        backward_fn = self._make_backward_fn(batch_idx, optimizer, opt_idx)
+        zero_grad_fn = self._make_zero_grad_fn(batch_idx, opt_idx, optimizer)
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:219]: Should the definition of these functions become part of `LightningClosure`

If the `LightningClosure` becomes the smallest unit of optimization, it means that it can be reused across loops of different semantic meanings.

In that case, It would be nice to be able to subclass the closure implementation without any changes to the loops

File: pytorch_lightning/loops/closure.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+@dataclass
+class ClosureResult:
+    closure_loss: Optional[Tensor]
+    result_collection: Optional[ResultCollection]
+
+    @property
+    def loss(self) -> Optional[Tensor]:
+        if self.closure_loss is not None:
+            return self.closure_loss.detach().clone()
```
Comment (carmocca) [inline pytorch_lightning/loops/closure.py:20]: Clone in `make_backward_fn` and here?

And what if you don't want to detach?

Comment (awaelchli) [reply pytorch_lightning/loops/closure.py:20]: Yes I know. This is just here ​to stick to how it was previously handled. I don't yet know how to handle this cleanly without breaking stuff.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 6 above)
Comment (yifuwang) [reply pytorch_lightning/loops/batch/training_batch_loop.py:219]: I think one benefit of having these outside of `LightningClosure` is we can creating different kind of closures (for handing different `training_step` signatures and optimization flows) without requiring `LightningClosure` to be aware of all of them.

File: pytorch_lightning/loops/closure.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/loops/closure.py:42]: > Do we really want a reference to this?
> This might introduce some unwanted longtime references...

so we definitely need a way to get the result out of the closure once it has run, therefore the reference here. This is the scenario:

```python
closure = Closure()
optimizer.step(closure)  # runs the closure, inside it will call model.training_step()
step_result = closre.result  # for further processing in our loops
```

Previously, it worked like this (pseudo-code):

```python
result = Result()  # empty result
closure = partial(loop.training_step_and_backward, result)
optimizier.step(closure)  # internally the closure will have a reference to result and call result.update(contents...). 
return result # here result contains stuff
```

> Usually we only need to create the closure once and can reuse it

So in the current design (and the previous one) the closure gets created on every step. 
This means the memory for the result here is getting freed once closure gets gc'd. 

However, if we want to build the closure object only once as you suggest, we may want to make the result property a get method that returns the result but also sets the self._reset to None. So how about this:

```python
def get_result(self):
    result = self._result
    self._result = None
    return result
```

File: pytorch_lightning/loops/closure.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/closure.py:23]: For why I did two classes:

The idea here was that the Closure abstract class defines the minimal interface for it to mimic a traditional closure that we can pass to the optimizer_step but that implements the notion of "getting a result back out of it". 

What runs inside the closure is specific to lightning's implementation, therefore the separate class LightningClosure.

For the naming, DefaultLightningClosure, I'm not fully convinced since we don't have this naming convention in Lightning.

Comment (ananthsub) [reply pytorch_lightning/loops/closure.py:41]: i asked about this here: https://github.com/PyTorchLightning/pytorch-lightning/pull/8700/files#r687310014

we are not consistent today on components prefixed with Lightning vs not. i lean towards not having the prefix because it's already under the lightning path pytorch_lightning.loops.closure

File: pytorch_lightning/loops/closure.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/loops/closure.py:60]: I like that a lot because that way we ensure to not have this reference forever. Could also be a property though, so that you can still do `closure.result` :)

File: pytorch_lightning/loops/closure.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loops/closure.py:76]: How about Closure and ClosureBase? Recently @ananthsub raised concern that we name too many internal components prefixed with Lightning*

File: pytorch_lightning/loops/closure.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/closure.py]: I renamed: Closure -> AbstractClosure, LightningClosure -> Closure

File: pytorch_lightning/loops/closure.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    variables without taking them explicitly as input. This has the benefit that a closure can be passed to an
+    object which later can call it like a function but without requiring to pass in any arguments.
+
+    This class provides a simple abstraction making the instance of this class callable like a function while capturing
+    the :class:`ClosureResult` and caching it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result = None
```
Comment (justusschock) [inline pytorch_lightning/loops/closure.py:55]:
```suggestion
        self._result: Optional[ClosureResult] = None
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/closure.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/closure.py]: Agree there, `get_result` is definitely a good choice.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 9]
```diff
@@ -199,7 +198,9 @@ def _run_optimization(
             if self.trainer.lightning_module.automatic_optimization:
                 self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
             else:
-                result = self._training_step(split_batch, batch_idx, opt_idx, self._hiddens)
+                closure()
+
+        result = closure.result
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:203]: Should we use closure.get_result() ?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 10]
```diff
@@ -211,37 +212,53 @@ def _run_optimization(
...
-        self,
-        split_batch: Any,
-        batch_idx: int,
-        opt_idx: int,
-        optimizer: Optimizer,
-        hiddens: Tensor,
-        return_result: AttributeDict,
-    ) -> Optional[Tensor]:
-        """Closure for training step and backward
+    def _make_closure(self, split_batch, batch_idx, opt_idx, optimizer, hiddens):
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:215]: docstring ?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 11]
```diff
@@ -211,37 +212,53 @@ def _run_optimization(
...
-            batch_idx: the index of the current batch
-            opt_idx: the index of the current optimizer
-            optimizer: the current optimizer
-            hiddens: the hidden state of the recurrent net
-            return_result: the storage of the trainstep results
-        """
+        closure = Closure(
+            step_fn=step_fn, backward_fn=backward_fn, zero_grad_fn=zero_grad_fn, profiler=self.trainer.profiler
+        )
+        return closure
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:223]:
```suggestion
        return Closure(
            step_fn=step_fn, backward_fn=backward_fn, zero_grad_fn=zero_grad_fn, profiler=self.trainer.profiler
        )
```

File: pytorch_lightning/loops/closure.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    This class provides a simple abstraction making the instance of this class callable like a function while capturing
+    the :class:`ClosureResult` and caching it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result: Optional[ClosureResult] = None
+
+    @property
+    def result(self) -> Optional[ClosureResult]:
```
Comment (tchaton) [inline pytorch_lightning/loops/closure.py:60]: I believe a function would be cleaner than a propery.

File: pytorch_lightning/loops/closure.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        self._zero_grad_fn = zero_grad_fn
+        self._profiler = PassThroughProfiler() if profiler is None else profiler
+
+    def closure(self, *args: Any, **kwargs: Any) -> Optional[ClosureResult]:
+        with self._profiler.profile("training_step_and_backward"):
+            output = self._step_fn()
+            output = ClosureResult(**output) if output else None
+
+            if output is None:
+                warning_cache.warn("training_step returned None. If this was on purpose, ignore this warning...")
```
Comment (tchaton) [inline pytorch_lightning/loops/closure.py:122]:
```suggestion
            output_step_fn = self._step_fn()
            output = ClosureResult(**output_step_fn) if output else None

            if output_step_fn is None:
                warning_cache.warn("training_step returned None. If this was on purpose, ignore this warning...")
```

File: pytorch_lightning/loops/closure.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        super().__init__()
+        self._step_fn = step_fn
+        self._backward_fn = backward_fn
+        self._zero_grad_fn = zero_grad_fn
+        self._profiler = PassThroughProfiler() if profiler is None else profiler
+
+    def closure(self, *args: Any, **kwargs: Any) -> Optional[ClosureResult]:
+        with self._profiler.profile("training_step_and_backward"):
+            output = self._step_fn()
+            output = ClosureResult(**output) if output else None
```
Comment (tchaton) [inline pytorch_lightning/loops/closure.py:119]: Not sure, but I believe ResultCollection is always provided from step_fn, making `ClosureResult`never empty.

Comment (awaelchli) [reply pytorch_lightning/loops/closure.py]: the LightningModule training_step can return None to skip an iteration, so that's how we are handling it here

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: addressed all your comments

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 15]
```diff
@@ -211,37 +213,79 @@ def _run_optimization(
...
+        def backward_fn(loss: Tensor):
+            self.backward(loss, optimizer, opt_idx)
+
+            # when in dev debugging track the losses
+            # TODO: remove dev debugger tracking loss history
+            self.trainer.dev_debugger.track_train_loss_history(batch_idx, loss)
+
+            # check if loss or model weights are nan
+            if self.trainer.terminate_on_nan:
+                check_finite_loss(self.trainer.lightning_module, loss)
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:281]: @awaelchli since this finite check was added here, can we remove `_process_closure_result`? (just below)

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:281]: yes, I think the way it is right now, it is redundant and can be removed. However, looking at it right now, the NaN checks are not the way they are supposed to be. 

The intention of terminate_on_nan is to check the loss BEFORE backward applies, and to check the weights after backwards. this must have been messed up in some of our loop refactors :( 

I will address this
